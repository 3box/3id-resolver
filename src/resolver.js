import { verifyJWT } from 'did-jwt'
import DidDocument from 'ipfs-did-document'
import base64url from 'base64url'
import { Resolver } from 'did-resolver'

const PUBKEY_IDS = ['signingKey', 'managementKey', 'encryptionKey']
const SUB_PUBKEY_IDS = ['subSigningKey', 'subEncryptionKey']

function validateDoc (doc) {
  let pubKeyIds = PUBKEY_IDS
  if (!doc || !doc.publicKey || !doc.authentication) {
    throw new Error('Not a valid 3ID')
  }
  if (doc.root) {
    pubKeyIds = SUB_PUBKEY_IDS
    if (!doc.space) throw new Error('Not a valid 3ID')
  }
  doc.publicKey.map(entry => {
    const id = entry.id.split('#')[1]
    if (!pubKeyIds.includes(id)) throw new Error('Not a valid 3ID')
  })
}

function encodeSection (data) {
  return base64url.encode(JSON.stringify(data))
}

async function verifyProof (subDoc, resolver) {
  const subSigningKey = subDoc.publicKey.find(entry => entry.id.includes(SUB_PUBKEY_IDS[0])).publicKeyHex
  const subEncryptionKey = subDoc.publicKey.find(entry => entry.id.includes(SUB_PUBKEY_IDS[1])).publicKeyBase64
  const payload = encodeSection({
    iat: null,
    subSigningKey,
    subEncryptionKey,
    space: subDoc.space,
    iss: subDoc.root
  })
  const header = encodeSection({ typ: 'JWT', alg: subDoc.proof.alg })
  const jwt = `${header}.${payload}.${subDoc.proof.signature}`
  await verifyJWT(jwt, { resolver })
}

function mergeDocuments (doc, subDoc) {
  subDoc.publicKey = doc.publicKey.concat(subDoc.publicKey)
  return subDoc
}

function getResolver (ipfs, { pin } = {}) {
  async function resolve (did, parsed) {
    async function _resolve (cid, isRoot = false) {
      let doc
      try {
        doc = await DidDocument.cidToDocument(ipfs, cid)
        validateDoc(doc)
        if (doc.root) {
          if (isRoot) throw new Error('Only one layer subDoc allowed')
          const rootDoc = await _resolve(doc.root.split(':')[2], true)
          await verifyProof(doc, localResolver)
          doc = mergeDocuments(rootDoc, doc)
        }
        if (pin) await ipfs.pin.add(cid)
      } catch (e) {
        try {
          if (pin) await ipfs.pin.rm(cid)
        } catch (e) {}
        throw new Error('Invalid 3ID')
      }
      return doc
    }
    return _resolve(parsed.id)
  }

  const resolveMethods = { '3': resolve }
  const localResolver = new Resolver(resolveMethods)
  return resolveMethods
}

module.exports = { getResolver }
