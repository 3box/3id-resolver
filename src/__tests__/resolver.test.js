import { Resolver } from 'did-resolver'
import base64url from 'base64url'
import { getResolver } from '../resolver'
import DidDocument from 'ipfs-did-document'
import { SimpleSigner, createJWT } from 'did-jwt'
import IPFS from 'ipfs'

const PRIV_KEY = '278a5de700e29faae8e40e366ec5012b5ec63d36ec77e8a2417154cc1d25383f'
const PUB_KEY = '03fdd57adec3d438ea237fe46b33ee1e016eda6b585c3e27ea66686c2ea5358479'

describe('3ID Resolver', () => {
  let ipfs, resolver

  beforeAll(async () => {
    ipfs = await IPFS.create({
      repo: './.ipfs',
      config: { Addresses: {}, Discovery: {}, Bootstrap: [] }
    })
  })

  beforeEach(() => {
    const threeIdResolver = getResolver(ipfs)
    resolver = new Resolver(threeIdResolver)
  })

  afterAll(async () => {
    await ipfs.stop()
  })

  describe('resolve 3ID', () => {
    let rootDID, rootCID

    describe('root 3ID', async () => {
      let doc

      beforeEach(async () => {
        doc = new DidDocument(ipfs, '3')
      })

      it('throws on invalid document', async () => {
        doc.addPublicKey('signingKey', 'Secp256k1VerificationKey2018', 'publicKeyHex', 'not a pub key')
        await doc.commit({ noTimestamp: true })
        await expect(resolver.resolve(doc.DID)).rejects.toEqual(new Error('Invalid 3ID'))
      })

      it('resolves valid document', async () => {
        doc.addPublicKey('signingKey', 'Secp256k1VerificationKey2018', 'publicKeyHex', PUB_KEY)
        doc.addPublicKey('encryptionKey', 'Curve25519EncryptionPublicKey', 'publicKeyBase64', 'fake encryptionKey')
        doc.addPublicKey('managementKey', 'Secp256k1VerificationKey2018', 'ethereumAddress', 'fake eth addr')
        doc.addAuthentication('Secp256k1SignatureAuthentication2018', 'signingKey')
        const cid = await doc.commit({ noTimestamp: true })
        const rawDoc = await DidDocument.cidToDocument(ipfs, cid)
        await expect(resolver.resolve(doc.DID)).resolves.toEqual(rawDoc)
        rootDID = doc.DID
        rootCID = cid
        await expect(ipfs.pin.ls(cid).next()).rejects.toMatchSnapshot()
      })

      it('should pin DID document if option given', async () => {
        const threeIdResolverPin = getResolver(ipfs, { pin: true })
        const resolver2 = new Resolver(threeIdResolverPin)
        await resolver2.resolve(rootDID)
        const pinnedCid = (await ipfs.pin.ls(rootCID).next()).value.cid
        expect(pinnedCid.toString()).toEqual(rootCID.toString())
        await ipfs.pin.rm(rootCID)
      })
    })

    describe('sub 3ID', () => {
      let doc, subDID, subCID

      beforeEach(async () => {
        doc = new DidDocument(ipfs, '3')
      })

      it('throws on invalid proof', async () => {
        doc.addPublicKey('subSigningKey', 'Secp256k1VerificationKey2018', 'publicKeyHex', 'fake subSigningKey')
        doc.addPublicKey('subEncryptionKey', 'Curve25519EncryptionPublicKey', 'publicKeyBase64', 'fake subEncryptionKey')
        doc.addAuthentication('Secp256k1SignatureAuthentication2018', 'subSigningKey')
        doc.addCustomProperty('space', 'a space name')
        doc.addCustomProperty('root', rootDID)
        const signature = base64url.encode('obviously invalid signature')
        doc.addCustomProperty('proof', { alg: 'ES256K', signature })
        await doc.commit({ noTimestamp: true })

        await expect(resolver.resolve(doc.DID)).rejects.toEqual(new Error('Invalid 3ID'))
      })

      it('resolves on valid sub document', async () => {
        const subSigningKey = 'fake subSigningKey'
        const subEncryptionKey = 'fake subEncryptionKey'
        const space = 'a space name'
        doc.addPublicKey('subSigningKey', 'Secp256k1VerificationKey2018', 'publicKeyHex', subSigningKey)
        doc.addPublicKey('subEncryptionKey', 'Curve25519EncryptionPublicKey', 'publicKeyBase64', subEncryptionKey)
        doc.addAuthentication('Secp256k1SignatureAuthentication2018', 'subSigningKey')
        doc.addCustomProperty('space', space)
        doc.addCustomProperty('root', rootDID)
        const payload = {
          subSigningKey,
          subEncryptionKey,
          space,
          iat: null
        }
        const signer = new SimpleSigner(PRIV_KEY)
        const jwt = await createJWT(payload, {
          issuer: rootDID,
          signer,
          alg: 'ES256K'
        })
        const signature = jwt.split('.')[2]
        doc.addCustomProperty('proof', { alg: 'ES256K', signature })
        const cid = await doc.commit({ noTimestamp: true })
        subDID = doc.DID
        subCID = cid

        await expect(resolver.resolve(doc.DID)).resolves.toMatchSnapshot()
        await expect(ipfs.pin.ls(cid).next()).rejects.toMatchSnapshot()
      })

      it('should pin DID document if option given', async () => {
        const threeIdResolverPin = getResolver(ipfs, { pin: true })
        const resolver2 = new Resolver(threeIdResolverPin)
        await resolver2.resolve(subDID)
        const pinnedCid1 = (await ipfs.pin.ls(rootCID).next()).value.cid
        expect(pinnedCid1.toString()).toEqual(rootCID.toString())
        const pinnedCid2 = (await ipfs.pin.ls(subCID).next()).value.cid
        expect(pinnedCid2.toString()).toEqual(subCID.toString())
      })
    })
  })
})
