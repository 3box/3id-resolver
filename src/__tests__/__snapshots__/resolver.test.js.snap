// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`3ID Resolver resolve 3ID root 3ID resolves valid document 1`] = `[Error: path 'bafyreid2wdjphtjzjpcnx2c2rmvrdahqdityows3or363ulushrgcxa4v4' is not pinned]`;

exports[`3ID Resolver resolve 3ID sub 3ID resolves on valid sub document 1`] = `
Object {
  "@context": "https://w3id.org/did/v1",
  "authentication": Array [
    Object {
      "publicKey": "did:3:bafyreidjnqr6o2ctd4tjn5tb6ohltzegrlhq6bxnfy5gucrlzwcg54wbw4#subSigningKey",
      "type": "Secp256k1SignatureAuthentication2018",
    },
  ],
  "id": "did:3:bafyreidjnqr6o2ctd4tjn5tb6ohltzegrlhq6bxnfy5gucrlzwcg54wbw4",
  "proof": Object {
    "alg": "ES256K",
    "signature": "IZ-NzQW0seb75WXJasUASe19sXw_xTBfmBc9L9GR-UwC-QaVF3x8v1uH6NPxYnWpxTNccEE99ZVuiA8xQqlNCA",
  },
  "publicKey": Array [
    Object {
      "id": "did:3:bafyreid2wdjphtjzjpcnx2c2rmvrdahqdityows3or363ulushrgcxa4v4#signingKey",
      "publicKeyHex": "03fdd57adec3d438ea237fe46b33ee1e016eda6b585c3e27ea66686c2ea5358479",
      "type": "Secp256k1VerificationKey2018",
    },
    Object {
      "id": "did:3:bafyreid2wdjphtjzjpcnx2c2rmvrdahqdityows3or363ulushrgcxa4v4#encryptionKey",
      "publicKeyBase64": "fake encryptionKey",
      "type": "Curve25519EncryptionPublicKey",
    },
    Object {
      "ethereumAddress": "fake eth addr",
      "id": "did:3:bafyreid2wdjphtjzjpcnx2c2rmvrdahqdityows3or363ulushrgcxa4v4#managementKey",
      "type": "Secp256k1VerificationKey2018",
    },
    Object {
      "id": "did:3:bafyreidjnqr6o2ctd4tjn5tb6ohltzegrlhq6bxnfy5gucrlzwcg54wbw4#subSigningKey",
      "publicKeyHex": "fake subSigningKey",
      "type": "Secp256k1VerificationKey2018",
    },
    Object {
      "id": "did:3:bafyreidjnqr6o2ctd4tjn5tb6ohltzegrlhq6bxnfy5gucrlzwcg54wbw4#subEncryptionKey",
      "publicKeyBase64": "fake subEncryptionKey",
      "type": "Curve25519EncryptionPublicKey",
    },
  ],
  "root": "did:3:bafyreid2wdjphtjzjpcnx2c2rmvrdahqdityows3or363ulushrgcxa4v4",
  "space": "a space name",
}
`;

exports[`3ID Resolver resolve 3ID sub 3ID resolves on valid sub document 2`] = `[Error: path 'bafyreidjnqr6o2ctd4tjn5tb6ohltzegrlhq6bxnfy5gucrlzwcg54wbw4' is not pinned]`;
